Add an init command for adding a UserSecretsId

liamdawson opened the pull request:
Adds an init command for adding a UserSecretsId

Addresses #166

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/DotNetTools?pullRequest=500) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/DotNetTools?pullRequest=500) before we can accept your contribution.<br/><br/>:x: liamdawson [sign now](https://cla.dotnetfoundation.org/aspnet/DotNetTools?pullRequest=500)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/DotNetTools?pullRequest=500) it.</sub>

liamdawson commented:
There are a couple of things I would like some help fixing appropriately:

1. Any test cases I'm missing? e.g. I feel I am missing test cases for invalid/missing project files for `init`, as they won't return the same output as they do for other commands right now.
2. As the above point, some error cases (invalid/missing project file) won't return output consistent with other subcommands - what's the best DRY way to solve that?
3. How should I indicate a command failure for https://github.com/aspnet/DotNetTools/pull/500/files#diff-0207ee8b1aa083ec78304322c8b620c7R65 ?
4. How do I add a localized error for "Project already has a UserSecretsId"? Adding it to the `resources.resx` didn't add the functions in the `designer.cs` that exist for the other resource entries.

natemcmaster commented:
Hi @liamdawson, thanks for doing this work! I'm going to help you merge this PR in. Unfortunately, we've missed the deadline for features in 2.2, so this PR will need to target 3.0 instead. I'm going to re-target your PR to the master branch and then review the code changes.

Thanks,
Nate

natemcmaster commented on src/dotnet-user-secrets/Internal/InitCommand.cs:
```diff
@@ -0,0 +1,99 @@
+                // TODO: i18n
+                context.Reporter.Error($"MSBuild project '{projectPath}' already contains a UserSecretsId.");
+                // TODO: correct error reporting approach
+                throw new NotImplementedException();
```
IMO tools should not fail if there action would be a no-op. They can just exit early. I think this code should be tweaked a little so we get the following behaviors.

Was `--id` given?  | Project contains `<UserSecretsId>` already? | Action
---|---|--
No | Yes | Message: "Project is already initialized", exit code 0
No | Yes | Generate GUID, add `<UserSecretsId>` element, exit code 0
Yes | No | Add `<UserSecretsId>` element, exit code 0
Yes | Yes | Overwrite `<UserSecretsId>` element, exit code 0

Reasons to fail:
* the ID specified with `--id` is invalid.
* Project file could not be found
* Project file is malformed (not proper XML)

natemcmaster commented on src/dotnet-user-secrets/Internal/InitCommand.cs:
```diff
@@ -0,0 +1,99 @@
+using Microsoft.Extensions.CommandLineUtils;
```
Two code formatting nit-picks and all new *.cs files:
* Can you add the copyright header?
* Can you sort using with System.* namespaces first? I wish VS enforced this better... :-/ https://github.com/aspnet/DotNetTools/blob/581b3160f422180b3411d84b4458c10128c8cacb/.editorconfig#L12-L14

natemcmaster commented on src/dotnet-user-secrets/Internal/InitCommand.cs:
```diff
@@ -0,0 +1,99 @@
+            // No valid property group, create a new one
+            if (propertyGroup == null)
+            {
+                propertyGroup = XElement.Parse(@"<PropertyGroup></PropertyGroup>");
```
Does this work? Parsing seems like a little much.
```
propertyGroup = new XElement("PropertyGroup");
```

liamdawson replied:
Right, I didn't actually try that because I was a bit superstitious about `XName`. Will take a look later.

natemcmaster commented on src/dotnet-user-secrets/Internal/InitCommand.cs:
```diff
@@ -0,0 +1,99 @@
+            }
+
+            // Add UserSecretsId element
+            var userSecretsElement = XElement.Parse(@"<UserSecretsId></UserSecretsId>");
```
Does this work?
```
var userSecretsElement = new XElement("UserSecretsId", newSecretsId);
```

natemcmaster commented on src/dotnet-user-secrets/Program.cs:
```diff
@@ -71,6 +71,12 @@ internal int RunInternal(params string[] args)
+            if (options.Command is InitCommand initCmd)
+            {
+                initCmd.Execute(new CommandContext(null, reporter, _console), _workingDirectory);
+                return 0;
```
In retrospec, Command.Execute should probably have returned `int` so the command can set an exit code. I'd be happy to also take a PR that made this change.

liamdawson replied:
Okay - I looked at that but ran into the fact the interface had the expectation. At first glance though, I don't know how I'd safely modify the interface to allow a status code return without causing a breaking change (but that's a separate conversation).

natemcmaster replied:
You have my blessing to make a breaking change on `ICommand`. For projects like this, the `ICommand` type is an internal implementation detail of these console tools. No one should be compiling an app against `dotnet-user-secrets.dll`.

liamdawson replied:
The `ICommand` in this instance is from CommandLineUtils, so not part of this tool - do you still want that change?

liamdawson commented:
Thanks for the thorough feedback @natemcmaster, I'll see to that soon.

Could I also get a pointer on how to add messages via `resources.resx`? I'll need to add a couple, and adding them in VS didn't seem to update the appropriate designer file (which says it is auto-generated).

Thanks again!

natemcmaster commented:
Re: resx files.

Edit the .resx file in VS (or an xml editor) and then run `dotnet msbuild -restore -t:Resx` from command line.

natemcmaster commented:
Sorry, missed a piece: run `dotnet msbuild -restore -t:Resx` _in the `src/dotnet-user-secrets/` project dir._

liamdawson commented:
I'll open another PR to cover the status code returns on `ICommand`, otherwise I believe I've addressed all your PR feedback (and have removed changes to korebuild and csproj files)

natemcmaster commented:
@liamdawson ACK saw your comments. I'll take a closer look at this in a few days. Currently working on time-sensitive issues. Please ping me on this PR  if I don't surface for air by Thurs. Nov. 1.

natemcmaster commented:
Thanks for making the requested changes. I'll bring this in once CI is green (I've just re-triggered a build).
